xc/device.ml

```diff
@@ -1430,6 +1431,111 @@ let can_surprise_remove ~xs (x: device) = Generic.can_surprise_remove ~xs x
 
 module Dm = struct
 
+  module QMP_Event = struct
+    let (pipe_r, pipe_w) = Unix.pipe ()
+    let channel_r = Unix.in_channel_of_descr pipe_r
+    let channel_w = Unix.out_channel_of_descr pipe_w
+
+    let epfd = Epoll.create1 [Epoll.EPOLL_CLOEXEC]
+
+    module Lookup = struct
+      let ftod, dtof = Hashtbl.create 16, Hashtbl.create 16
+      let add fd domid =
+        Hashtbl.replace ftod fd domid;
+        Hashtbl.replace dtof domid fd
+      let remove fd domid =
+        Hashtbl.remove ftod fd;
+        Hashtbl.remove dtof domid
+      let domid_of fd = try Some (Hashtbl.find ftod fd) with Not_found -> None
+      let fd_of domid = try Some (Hashtbl.find dtof domid) with Not_found -> None
+    end
+
+    module Monitor = struct
+      let create () = Epoll.create1 [Epoll.EPOLL_CLOEXEC]
+      let add m fd = Epoll.ctl m Epoll.EPOLL_CTL_ADD fd [Epoll.EPOLLIN]
+      let remove m fd = Epoll.ctl m Epoll.EPOLL_CTL_DEL fd [Epoll.EPOLLIN]
+      let wait m = Epoll.wait m (Hashtbl.length Lookup.dtof + 1) (-1)
+      let with_event ~ok ~err = function | Epoll.EPOLLIN -> ok() | e -> err e
+    end
+    let m = Monitor.create ()
+
+    let monitor_path domid = Printf.sprintf "/var/run/xen/qmp-event-%d" domid
+    let debug_exn msg e = debug "%s: %s" msg (Printexc.to_string e)
+
+    let remove domid =
+      try
+        match Lookup.fd_of domid with
+        | None -> ()
+        | Some c -> Lookup.remove c domid;
+          Monitor.remove m (Qmp_protocol.to_fd c);
+          Qmp_protocol.close c;
+          debug "Removed QMP Event fd for domain %d" domid
+      with e -> debug_exn (Printf.sprintf "Get exception trying remove qmp on domain-%d" domid) e
+
+    let add domid =
+      try
+        match Lookup.fd_of domid with
+        | Some _ -> ()
+        | None -> let c = Qmp_protocol.connect (monitor_path domid) in
+          Qmp_protocol.negotiate c;
+          Lookup.add c domid;
+          Monitor.add m (Qmp_protocol.to_fd c);
+          debug "Added QMP Event fd for domain %d" domid
+      with e ->
+        debug_exn (Printf.sprintf "QMP domain-%d: negotiation failed: removing socket" domid) e;
+        remove domid
+
+    let wakeup domid =
+        output_string channel_w (Printf.sprintf "%d\n" domid);
+        flush channel_w
```

Comment:
~output_string channel_w (Printf.sprintf "%d\n" domid);~
~flush channel_w~
add domid;
(* write single byte to wake up Monitor.wait *)
Unix.write pipe_w " " 0 1